Add support for getting data from morph.io scrapers

chrismytton opened the pull request:
# What does this do?

This adds ~~3~~ 2 new methods for getting data from a remote morph scraper, `MorphScraper::Database#data`, ~~`MorphScraper::Database#table`~~ and `MorphScraper::Database#query`. All of these methods return an array of hashes representing the records from the remote scraper's database.

``` ruby
db = MorphScraper::Database.new('tmtmtmtm/malta-parliament')

# Equivalent to 'SELECT * FROM data'
db.data

# Equivalent to 'SELECT * FROM terms'
db.data(:terms)

# Or just execute arbitrary queries
db.query('SELECT *, 5 as term FROM data')
```
# Why was this needed?

This gem already [has a method](https://github.com/everypolitician/morph_scraper-database/blob/bcfd6fb77d288d2e59ca5405fea9f561f1ba6db3/lib/morph_scraper/database.rb#L14-L21) for copying the database from another morph scraper into the current scraper's `data.sqlite`, but it completely overwrites what's in the current scrapers database. This is OK if you've just created a new scraper and want the new scraper's database to be identical to the old one, but falls down for anything more complicated than that.

To solve this problem this pull request adds the methods mentioned above to give you full programatic access to the contents of a morph scraper. This can be used by a scraper to import the partial contents of another scraper. It can also potentially be used as part of the everypolitician-data build process to make it easier to work with remote scrapers.
# Implementation notes

This is built on top of [Sequel](https://github.com/jeremyevans/sequel), so most of the heavy lifting is done by that gem, with this gem just shuttling the data around and providing some convenience methods for accessing the data.

chrismytton committed:
Create a Tempfile for the database

This will allow us to open the remote database without needing to write
it out to the scraper, we can just interface with it directly from the
Tempfile.

chrismytton committed:
Add sequel and sqlite3 dependencies

chrismytton committed:
Move with_tmp_dir helper into test_helper

chrismytton committed:
Disable test warnings for now

WebMock gives lots of warnings because it's monkey patching various
things. Turn the warnings off for now so we can focus on the tests.

chrismytton committed:
Add pry to test_helper.rb

chrismytton committed:
Add a create_database helper for the tests

This generates a temporary sqlite database on the fly for use in the
tests.

chrismytton committed:
Add Database#data method

This returns the contents of the `data` table as an array.

chrismytton committed:
Add Database#table method

This allows you to access any table in the remote scraper's database as
an Array of Hashes.

chrismytton committed:
Add Database#query method

This allows you to run an arbitrary select query and get the result back
as an Array of Hashes.

chrismytton committed:
Add support for Ruby 2.1 and 2.2

These version of Ruby require a `basename` argument to the `Tempfile`
constructor, so provide them with one to make the gem work on Rubies
less than 2.3.

chrismytton commented:
@tmtmtmtm 👀

tmtmtmtm reviewed (changes requested):
Woot! I'm not totally convinced by the naming of the `table` method though. I'm not sure what I would have expected such a method to actually be, but having it return the _contents_ of that table would probably have been quite far down that list.

I'm not totally sure what would be better though. Did you consider having `data` take an optional argument for which table to read from?

PS: You also seem to have forgotten to update the README…

chrismytton committed:
Drop the confusingly named #table method

It wasn't clear what this method was supposed to do from the name.
Futhermore there is already a method that returns the contents of a
table, `#data`. So I've dropped the `#table` method an added an optional
argument to the `#data` method which is the name of the table you want
to retrieve data from, defaulting to `:data`.

chrismytton committed:
Add section for replacing local database to README

This is so that I can add another section in the next commit which will
explain how to get a subset of the data from a scraper.

chrismytton committed:
Add details about getting data from a scraper to README

chrismytton commented:
> I'm not totally sure what would be better though. Did you consider having data take an optional argument for which table to read from?

Ah, I hadn't considered that because in my head the method was called `#data` because it was returning rows from the `data` table. However on further consideration I think that having a general purpose `#data` method that takes an optional table name and defaults to the `data` table makes sense. I've made that change in https://github.com/everypolitician/morph_scraper-database/pull/1/commits/c90fc47e9ae8e41ae080eab1a45f9639dbb2a564.

> PS: You also seem to have forgotten to update the README…

😳  Done in dfc323b765b543c62f6956b168810d38cd550753

@tmtmtmtm 👀

tmtmtmtm commented on README.md:
```diff
@@ -37,7 +39,7 @@ MorphScraper::Database.new('tmtmtmtm/malta-parliament').write(force: true)
 
 **Note**: The above code will overwrite the database of the _current_ scraper with the contents of the _named_ scraper's database **every single time** this code is run. You might want to make this code conditional on an environment variable or remove it once you've used it, otherwise it will overwrite your database on each run and you can potentially loose data.
```
NB: not from this change, but there's a typo here: 'loose' should be 'lose'

chrismytton commented:
Just discussed this in Slack with @tmtmtmtm and decided that adding a dependency on `sequel` _and_ `sqlite3` is probably a bit extreme for our use case.

I think a better course of action will be to execute the queries against morph's API (#2) which is more in keeping with our use pattern, where we generally just do one query against one scraper. This also means we won't need to add any extra dependencies. Then we can come back to this pull request and see if we _actually_ want to add support for using a local SQLite database.

chrismytton commented:
The simpler version that use the Morph JSON API was implemented in #3, so I'm going to close this.

We might want to revisit the concept of having a local database and querying it with Sequel in the future, but it will almost certainly look different to this pull request.
